lib/spack/spack/test/llnl/util/filesystem.py

```diff
@@ -908,3 +908,67 @@ def test_find_first_file(tmpdir, bfs_depth):
 
     # Should find first dir
     assert os.path.samefile(fs.find_first(root, "a", bfs_depth=bfs_depth), os.path.join(root, "a"))
+
+
+def test_rename_dest_exists(tmpdir):
+
+    def setup_test_files(root):
+        a = tmpdir.join(root, "file1")
+        b = tmpdir.join(root, "file2")
+        fs.touch(a)
+        fs.touch(b)
+        return a, b
+
+    def setup_test_dirs(root):
+        a = tmpdir.join(root, "a")
+        b = tmpdir.join(root, "b")
+        fs.mkdirp(a)
+        fs.mkdirp(b)
+        return a, b
+
+    # test standard behavior of rename
+    # smoke test
+    a, b = setup_test_files("a")
```

Comment:
A couple requests:

* A function that more-clearly restores a "clean" working state of the root would be good (and call it before each sub test) - this could be achieved by parameterizing the test (although pushing the various differences into pytest parameters can be annoying, so other approaches are good too)
  * I think the main thing that would be useful is (a) call this before each sub test and (b) make sure to delete everything in the tmpdir and re-instantiate it from scratch
* If the files have content, then you can make sure for example that when renaming x to y, that y now contains the content that x had
  * likewise if renaming symlink x to symlink y, the "new" y should now point to whatever x was pointing to